### ISSUE TITLE
Add pagination on results

### ISSUE THREAD

**Issue by pierresouchay** [hashicorp/vault#10887]
In our Vault instances, we have sometimes several thousands of tokens and secrets.

When we have too many instances (eg: `vault list auth/token/accessors`), then the response is:

```
Error listing auth/token/accessors/: context deadline exceeded
```

Augmenting the timeout is not a solution, we need paging for that.

Do you think you might consider adding paging to all methods supporting returning an unknown number of results?

A few examples of people having similar issues:
* https://github.com/hashicorp/vault/issues/8598
* https://discuss.hashicorp.com/t/is-it-possible-to-paginate-auth-token-accessors/14857

**Comment by emas80** [comment]
I have the same problem...
I have to revoke some 7k tokens which do not have a ttl, and I don't know how I can achieve that if I am unable to list the tokens or the accessors.

**Comment by pierresouchay** [comment]
@emas80 IIRC, we ended with deleting the tokens on our Consul KV by ourselves...

**Comment by jpedrick** [comment]
Hi @hellobontempo , moving this conversation... from: https://github.com/hashicorp/vault/issues/7079#issuecomment-1593624960

Thanks for the update on that. What does "Discovery" mean?

One thing I'd note, the customer's workaround here in #10887 sounds like HCV is fundamentally broken at scale. If the tokens they had were valid, they would not have been able to just delete them from the backend.

I hope Hashicorp internally recognizes the priority of this issue. I haven't gotten a clear signal on what the priority or timeline of resolving this is yet. As an open source user, I am unable to decide if I should attempt contributing to solving this issue or if someone within Hashicorp is actively working on it.

If Hashicorp would like external contributions, they should bring the design discussion to a public forum. If not, they should inform the community what the priority and rough timeline is. As is, since this issue has sat idle for over 2 years, I would assume it's not actually being worked on.

**Comment by hsimon-hashicorp** [comment]
Hi @jpedrick, thanks for sharing your thoughts with us. We appreciate that our practitioners and contributors are very enthusiastic about Vault and share HashiCorp’s goals of making it a “best in class” product. I’d like to take a moment to remind you of [HashiCorp’s Community Guidelines](https://www.hashicorp.com/community-guidelines), because even though we can appreciate that these are things our community feels strongly about, we like to make sure everyone is being respectful in these discussions. As to your questions, “Discovery” means just that - the team is taking some time to perform some research into the issue as well as potential solutions and any downfalls of said potential solutions.

The nature of our company is such that we have open source contributors, and a large team of employee engineers. My job is to help facilitate meaningful and respectful collaboration between these two groups of people. I understand it can be confusing or frustrating when there’s a lack of visibility into what the engineering team is currently doing, but I assure you we are always looking for ways to improve the product for all of our users, and that includes our HCP Vault users.

**Comment by maxb** [comment]
I was reading the above exchange between @jpedrick and @hsimon-hashicorp over dinner yesterday ... but now I've found time to react, it appears that @jpedrick has deleted his comment (I assume? Since it's gone?) after being reminded of the Community Guidelines ... :-(

I think that's rather unfortunate, as whilst it contained one notably undiplomatically phrased sentence, it was for the most part a valid statement of actual concerns.

Summarizing what was said:

* The issue reported here is a serious problem when attempting to operate Vault at scale. (Since it prevents admins from auditing the active tokens in their clusters.)

* Despite the issue being pretty important, it got no reaction from HashiCorp (at least, so far as is publically visible) whilst being open for 2 years, which is a bit unsettling to potential Vault adopters. (Even now, the feedback acknowledging it is now being thought about, is a brief comment only, in a **different** issue: https://github.com/hashicorp/vault/issues/7079#issuecomment-1593624960)

* As a potential open-source contributor, it would be nice to know whether HashiCorp already has a design in mind, and if not, whether HashiCorp is open to entering into a conversation about accepting an external contributor's PR for this issue. Whilst there's clearly a need for HashiCorp to approve any proposed design, API pagination is not exactly rocket science, and a contributor could potentially make this happen sooner.

**Comment by raskchanky** [comment]
@maxb I agree with you that pagination itself isn't rocket science, but Vault's design makes pagination a little more tricky than the average web app. We do already have a design in mind for how this might work, but it builds on top of a few new features that haven't been built yet. Once those are done, adding pagination will be a lot easier than it is today. We definitely appreciate everyone's patience while we work through all this. Sometimes balancing priorities is tough.

**Comment by hsimon-hashicorp** [comment]
@maxb it is unfortunate that he deleted his comment. Vault is a complex product containing many moving parts - as I'm sure you're aware! - and that level of complexity means that not every request for an enhancement is going to be as easy as accepting a community PR. Often, there are months of work, sometimes over one or more releases, required to set up the groundwork. As a mission-critical security product for many practitioners, it behooves us to proceed with caution.

As I said earlier, the lack of visibility into these planned features or enhancements can be frustrating for the community, and I do understand that, and we appreciate folks' patience.

**Comment by maxb** [comment]
> We do already have a design in mind for how this might work, but it builds on top of a few new features that haven't been built yet. Once those are done, adding pagination will be a lot easier than it is today.

Are you able to share anything about those foundational features on the path to pagination?

I'm mainly curious, because the code seems like it's already in a good place to start implementing pagination, to me.

**Comment by ellisroll-b** [comment]
We are generally happy with vault, and were considering putting A LOT of secrets in it (millions). We do have complex path (tree) but still would have thousands to 100's of thousands under a path. We have several use cases where we need to remove all under a path, or matching patterns (which generally would require enumeration).  Given the lack of pagination, we will have to find another repo to put keys in (which will likely not be as secure or nice, but we **have** to scale).

**Comment by heatherezell** [comment]
Hello folks! We appreciate the feedback here. Our product managers are looking for folks willing to chat in deeper detail (ie via Zoom). Please email me (my username here, at hashicorp dot com) if you're willing to set up a deeper-dive conversation around your needs.

**Comment by pierresouchay** [comment]
> Hello folks! We appreciate the feedback here. Our product managers are looking for folks willing to chat in deeper detail (ie via Zoom). Please email me (my username here, at hashicorp dot com) if you're willing to set up a deeper-dive conversation around your needs.

@heatherezell I sent you an email to your username in GitHub at hashicorp com, but I received a delivery failed notification.

**Comment by ellisroll-b** [comment]
Thanks for the offer @heatherezell. However we need cloud scale, and OSS stability.  Given how long this has been an issue, we have chosen to use Vault only for small scale things, (and have moved to other solutions for large scale).  We would also be hesitant to engage further until after we what happens post the acquisition; as the license agreement change, and then IBM acquisition, has given us pause to continue in the use of multiple Hashicorp products.

**Comment by heatherezell** [comment]
> Thanks for the offer @heatherezell. However we need cloud scale, and OSS stability. Given how long this has been an issue, we have chosen to use Vault only for small scale things, (and have moved to other solutions for large scale). We would also be hesitant to engage further until after we what happens post the acquisition; as the license agreement change, and then IBM acquisition, has given us pause to continue in the use of multiple Hashicorp products.

I appreciate your honesty and candor! This is still useful information, and I'll bring it back to our product team. Thank you!

**Comment by pierresouchay** [comment]
@heatherezell while being a very valued hashicorp contributor (I was the first WW external contributor to Consul for a while, scaling a lot the product) and having privileged relationship with the company (commercial and OSS), I tried several times to move on this topic, and nobody cared. We even submitted patches for it 

As explained in comments above, this is critical for scale, listing without paging make it a toy that does not work at scale
